neurovault/apps/statmaps/tasks.py

```diff
@@ -24,9 +29,84 @@ def generate_glassbrain_image(image_pk):
     plt.close('all')
     return png_img_path
 
+# IMAGE TRANSFORMATION ################################################################################
+
+# Save 4mm, brain masked image vector in pkl file in image folder
+@shared_task
+def save_resampled_transformation_single(pk1, resample_dim=[4, 4, 4]):
+    from neurovault.apps.statmaps.models import Image
+    from nilearn.image import resample_img
+    import neurovault
+    img = get_object_or_404(Image, pk=pk1)
+    nii_obj = nib.load(img.file.path)
+    
+    # If a standard already exists with the voxel dimension, we use that to save time
+    standard = os.path.abspath(os.path.join(neurovault.settings.BASE_DIR,
+                                                  "static","anatomical",
+                                                  "MNI152_%smm.nii.gz" % resample_dim[0]))
+    if not os.path.exists(standard):
+      standard = os.path.abspath(os.path.join(neurovault.settings.BASE_DIR,"static","anatomical","MNI152.nii.gz"))
+    standard_brain = nib.load(standard)
+    
+    # Resample both image and mask - resampling will be skipped if affines already equivalent
+    img_resamp, ref_resamp = resample_images_ref(images=[nii_obj], 
+                                                 reference=standard_brain, 
+                                                 interpolation="continuous",
+                                                 resample_dim=resample_dim)
+
+    # Mask the image, and save pickle image folder 
+    # (this is the same procedure used to produce the atlas vector that will be used for scatterplot)
+    image_vector = img_resamp[0].get_data()[ref_resamp.get_data()!=0]
+
+    pkl_resamp_name = "transform_%smm_%s.pkl" %(resample_dim[0],img.pk)
+    pkl_img_path = os.path.join(os.path.split(img.file.path)[0], pkl_resamp_name)
+    if os.path.exists(pkl_img_path):
+        os.unlink(pkl_img_path)
+    pickle.dump(image_vector,open(pkl_img_path,"wb"))
+
+    # Update the image "transform" field with the pkl_img_path
+    img.transform = pkl_img_path
+    img.save()
+
+    return pkl_img_path
+
+
+# SIMILARITY CALCULATION ##############################################################################
 
 @shared_task
-def save_voxelwise_pearson_similarity(pk1, pk2, resample_dim=[4, 4, 4]):
+def run_voxelwise_pearson_similarity(pk1):
+
+    # Make sure we have a transform for pk in question
+    from neurovault.apps.statmaps.models import Image, Comparison
+    image1 = get_object_or_404(Image, pk=pk1)
+    if image1.transform is None:
```

Comment:
ok, I implemented this - now we have a similar problem, but an integrity error with the comparison itself, [happening here](https://github.com/vsoch/NeuroVault/blob/update/image_comparison/neurovault/apps/statmaps/tasks.py#L139):

      try:
            compare = Comparison.objects.get(image1=image1, image2=image2,
                                         similarity_metric=pearson_metric)
      except Comparison.DoesNotExist:
            compare = Comparison(image1=image1, image2=image2,  similarity_metric=pearson_metric)
      compare.similarity_score = pearson_score
      compare.save()

We have the situation that when new images, let's say pk1 and pk2 are added (via save()):

- both get called run_voxelwise_pearson_similarity(pk)
- For each of image1 and image2, pearson scores are calculated for the image vs all others with sorted ids, after a check to see the comparison doesn't exist. There is a second check within the function, just before saving.
- since these jobs are run at the same time, the initial check will not find the comparison for either image1 or image2, and both jobs to calculate the pearsons run
- both jobs generate the transforms atomically (no trouble with this so far)
- within the function, when image1 checks to see if the comparison exists, it does not. It adds to the image object and then tries to save.
- within that time, image2 has already saved the comparison for the set, and we get the integrity error:

      IntegrityError: duplicate key value violates unique constraint    "statmaps_comparison_image1_id_2aaf904faeb6c3c5_uniq"
      DETAIL:  Key (image1_id, image2_id)=(370, 371) already exists.

My intuition is that the Comparison model needs some kind of custom save() function, but I wanted to get your thoughts on this first.